Tapscript new opcodes

sanket1729 opened the pull request:
Please refer to the new spec here. https://github.com/ElementsProject/elements/pull/1019

apoelstra commented on src/crypto/sha256.cpp:
```diff
@@ -705,6 +705,60 @@ void CSHA256::Midstate(unsigned char hash[OUTPUT_SIZE], uint64_t* len, unsigned
+    return true;
+}
+
+bool CSHA256::SafeWrite(const unsigned char* data, size_t len) {
```
weird indentation in this function.

Also cc @roconnor-blockstream is this function really necessary? We always provide the size of an in-memory buffer as `len`, which will never approach 2^61.

roconnor-blockstream replied:
@apoelstra I'm not sure you understand the nature of this function.  The sha256 opcodes read the sha256 context from the stack, and thus is user-generated values.  There is no data abstraction, and thus the user can load up a SHA256 context that pretends that it has been filled, or nearly filled with SHA256's maximum number of bytes.  How we choose to deal with an overflow, even if the overflow is based on a lie, is a consensus critical decision.

apoelstra replied:
Thanks! Yes, I missed this when I first read the function, then didn't revisit after reading how the streaming SHA opcodes worked.

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -60,6 +60,38 @@ static inline void popstack(std::vector<valtype>& stack)
+    stack.push_back(vchAssetPref);
+}
+
+// Review: Not sure if this should inline or not? Leave it to the compiler??
```
Might as well inline both of these for consistency's sake. But it really doesn't matter.

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -60,6 +60,38 @@ static inline void popstack(std::vector<valtype>& stack)
 }
 
+static inline void pushasset(std::vector<valtype>& stack, const CConfidentialAsset& asset)
+{
```
Can you `assert` here that `!asset.IsNull()`? (I believe this is enforced for any transaction input...although I wonder what happens for e.g. coinbase transactions?)

apoelstra replied:
Or you could just push `0` onto the stack in case the asset is null, and not worry about whether this is actually possible.

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1680,6 +1712,100 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                            valtype vchPrevTxid;
+                            vchPrevTxid.insert(vchPrevTxid.begin(), inp.prevout.hash.begin(), inp.prevout.hash.begin() + 32);
+                            valtype vchPrevVout;
+                            auto vout_le = htole32(inp.prevout.n);
```
for the avoidance of doubt can you do `uint32_t` rather than `auto` here?

I'm also a little uneasy about using `htole32` directly here but I see it's used already in `OP_DETERMINISTICRANDOM` so ok.

apoelstra replied:
Ah I see `htole32` is encapsulated in the next commit into its own `push4_le` method. Ok, I'm happy with that.

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1680,6 +1712,100 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                                stack.push_back(vchAssetBlindingNonce);
+                            } else { // No issuance
+                                stack.push_back(vchFalse);
+                            }
```
It occurs to me that there is no way to get the actual asset ID (though maybe you can gin it up using the streaming hash opcodes..). Worth thinking about in a future softfork, I don't want to extend the scope of this one too much.

roconnor-blockstream replied:
I think we can compute the asset ID with our CAT and SHA256.  If not, we can add new opcodes at a later date.

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1806,6 +1809,129 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                        case 0: // Version
+                        {
+                            // This does an inplicit conversion from version(int32_t) to (uint32_t)
+                            push4_le(stack, (uint32_t) checker.GetTxVersion());
```
spelling: `implicit`.

Also it's not implicit, because you have an explicit cast :)

roconnor-blockstream replied:
I believe static_casts are preferred in C++.

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -2535,6 +2661,12 @@ unsigned int GenericTransactionSignatureChecker<T>::GetTxVersion() const
+{
+    return ::GetSerializeSize(*txTo, PROTOCOL_VERSION | SERIALIZE_TRANSACTION_NO_WITNESS) * (WITNESS_SCALE_FACTOR - 1) + ::GetSerializeSize(*txTo, PROTOCOL_VERSION);
+}
+
```
looks like this is just copied from `GetTranasactionWeight` in src/validation.h .... can you directly call that method rather than reimplementing it?

sanket1729 replied:
I could not do it because the function is defined on a concrete type `CTransaction` and we are using a generic type name here.

apoelstra replied:
Can you try casting by passing `CTransaction(*txTo)`?

apoelstra replied:
Hmm, the compiler should let you do this

sanket1729 replied:
Yep, it does.

sanket1729 replied:
The more I think the *txTo might be creating a copy.

apoelstra replied:
No, `GetSerializeSize` takes its input by reference. There is no copy.

apoelstra commented on test/functional/feature_tapscript_opcodes.py:
```diff
@@ -142,6 +235,64 @@ def run_test(self):
+        self.tapscript_satisfy_test(CScript([OP_0, OP_0, OP_INSPECTINPUT, b'\x00', OP_EQUAL]))
+        # 1b. Add a pegin (Test pegin input must be 0x40)
+        self.tapscript_satisfy_test(CScript([OP_0, OP_0, OP_INSPECTINPUT, b'\x40', OP_EQUAL]), add_pegin=True)
+        # 1c. Add peing (The non-pegin input must be push 0x00)
```
spelling `pegin`

apoelstra commented on test/functional/feature_tapscript_opcodes.py:
```diff
@@ -6,10 +6,11 @@
 #
 # Test for taproot sighash algorithm with pegins and issuances
 
-from test_framework.key import compute_xonly_pubkey, generate_privkey, sign_schnorr, tweak_add_privkey, tweak_add_pubkey, verify_schnorr
```
This test does not pass in the commit that it is added 14d5cac0acc486397c9330696f313405ff92cbbc

apoelstra replied:
also can you `chmod +x` the file?

apoelstra replied:
now added in 1dfe55190b40a33bd4d4217ad86e6b8c76621b34 .. but still does not pass

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -2714,8 +2712,26 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
     }
 
     // Run the script interpreter.
+    // If serror is nullprt, set it to freshly created instance
```
spelling `nullptr`

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1800,11 +1807,8 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
-                            // Alternatively, this can be turned into a OP_NOP, but restricts the use of this
-                            // opcode to read-only type.
-                            return set_error(serror, SCRIPT_ERR_INTROSPECT_INDEX_OUT_OF_BOUNDS);
+                            // Operate as OP_SUCCESS for all undefined intospection positions
```
spelling `introspection` (same thing in a couple places below)

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -2714,8 +2712,26 @@ static bool ExecuteWitnessScript(const Span<const valtype>& stack_span, const CS
+    }
     if (!EvalScript(stack, scriptPubKey, flags, checker, sigversion, execdata, serror)) return false;
 
+    // Set script success for early return. This skips the checks for CLEANSTACK, SCRIPT_ERR_EVAL_FALSE check.
```
Also, I don't think you need to set `serror` back to null. I think you can just do
```
ScriptError _serror;
if (serror == nullptr) serror = &_serror;  // ELEMENTS: we need this error value even if the caller doesn't
```
and then leave the rest of the existing logic alone.
￼

apoelstra replied:
We never return this pointer. We return a boolean.

apoelstra commented on src/script/script.cpp:
```diff
@@ -156,6 +156,25 @@ std::string GetOpName(opcodetype opcode)
     case OP_INSPECTOUTPUT          : return "OP_INSPECTOUTPUT";
     case OP_INSPECTTX              : return "OP_INSPECTTX";
 
+    // 64 bit LE arithmatic opcodes
```
spelling: `arithmetic` (I had to look this one up :)). Appears in a few places and also in the commit message.

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1929,6 +1937,178 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                            // fallthrough (a - b == a + (-b)). Negation should not change any overflow logic
+                        case OP_ADD64:
+                            if ((a > 0 && b > std::numeric_limits<int64_t>::max() - a) ||
+                                (a < 0 && b < std::numeric_limits<int64_t>::min() - a))
```
In theory these limits are implementation defined.

I would prefer you use the constants `0xFFFFFFFFFFFFFFFF` and `-0x10000000000000000`.

cc @roconnor-blockstream what do you think?

sanket1729 replied:
These constants are also used in bench tests. AWaiting response from Russell before changing.

roconnor-blockstream replied:
How are they implementation defined?  Are you worried about non-two's complement?

apoelstra replied:
I am worried about non-two's-complement, yes. And I say "implementation defined" because I could not find a citation saying that these constants should have specific values.

I'd appreciate a static assert that these values are what we expect, but I don't feel too strongly about this. In practice I doubt that our software (or Bitcoin) will work properly on a twos-complement system, and I'm not aware of any real-world users with such systtems.

roconnor-blockstream replied:
Static_asserts are still needed.

sanket1729 replied:
Added at the top of the file

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1929,6 +1937,178 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                }
+                break;
+
+                case OP_SCIPTNUMTOLE64:
```
spelling `SCRIPTNUMTOLE64` (and a few other instances of `SCIPT`)

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1929,6 +1937,178 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                                return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+                            int64_t offset = CScriptNum(stacktop(-2), fRequireMinimal).getint();
+                            if (offset < 0 || offset > 63 || a < 0)
+                                return set_error(serror, SCRIPT_ERR_ARITHEMATIC64);
```
spelling `ARITHMETIC`

apoelstra commented on src/script/script_error.cpp:
```diff
@@ -119,6 +119,10 @@ std::string ScriptErrorString(const ScriptError serror)
         case SCRIPT_ERR_INTROSPECT_INDEX_OUT_OF_BOUNDS:
             return "Introspection index out of bounds";
+        case SCRIPT_ERR_EXPECTED_8BYTES:
+            return "Arethmatic opcodes expect 8 bytes operands";
```
spelling `Arithmetic`

apoelstra commented on src/script/script_error.cpp:
```diff
@@ -119,6 +119,10 @@ std::string ScriptErrorString(const ScriptError serror)
             return "Introspection index out of bounds";
+        case SCRIPT_ERR_EXPECTED_8BYTES:
+            return "Arethmatic opcodes expect 8 bytes operands";
+        case SCRIPT_ERR_ARITHEMATIC64:
```
and here

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1643,29 +1690,33 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                         popstack(stack);
-
-                    if (!fSuccess)
-                        return set_error(serror, SCRIPT_ERR_CHECKSIGVERIFY);
```
I think this line was not supposed to be deleted (or it was supposed to be moved)

sanket1729 replied:
Good catch. I think this code always failed if provided with a non-empty sig `CHECKSIGSTACK`. Do we want to change the behavior so that we allow empty sigs to fail without aborting? Or stick with the existing behavior where CHECKSIGFROMSTACK cannot be failed. In which case, there is little difference between CHECKSIGFROMSTACK and the verify version.

sanket1729 replied:
My thinking is that this was not intended in the original implementation and we allow CHECKSIGFROMSTACK to fail.

apoelstra replied:
We cannot fix this bug without a hardfork :)

sanket1729 replied:
Yeah, I was intending to say that we should fix it for tapscript

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1937,6 +1988,56 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    valtype& vchRes = stacktop(-3);
+
+                    CPubKey pk(vchGenerator);
+                    CPubKey res(vchRes);
```
I think we may want to restrict to compressed keys here. Thoughts?

apoelstra replied:
If you disagree, the test should include uncompressed and hybrid keys

sanket1729 replied:
Agreed. I wanted to restrict to compressed keys, but this is an oversight.

sanket1729 reviewed (commented):
Removed the WIP from title. Ready for another review round.

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -2385,6 +2385,24 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(const CScriptNum& nSeq
+template <class T>
+const CTransaction* GenericTransactionSignatureChecker<T>::GetTransaction() const
+{
+    return (CTransaction*)(txTo);
```
Can you use `static_cast<CTransaction*>` rather than a C-style cast? This will get some compile-time checking that the cast makes sense.

You can also use the CTransaction constructor, like `return CTransaction(*txTo);` although I expect this will make a copy.

sanket1729 replied:
Using static_cast

sanket1729 replied:
Switched back to original implementation that copies the tx weight calculation from `src/consensus/validation.cpp`.

apoelstra replied:
Cool. Makes sense, given offline discussion about what our options are here.

apoelstra commented on test/functional/feature_tapscript_opcodes.py:
```diff
@@ -379,5 +380,126 @@ def run_test(self):
+        def check_rshift(a, l, res, fail=False):
+            self.tapscript_satisfy_test(CScript([l, le8(a), OP_RSHIFT64 , le8(res, False), OP_EQUAL]), fail=fail)
+        # Arithematic opcodes
+        self.log.info("Check Arethmetic opcodes")
```
spelling :) `arithmetic`

apoelstra replied:
Ah, I see this is fixed two commits later

sanket1729 replied:
Oops, I fixed this in the wrong commit. Changing it

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1937,6 +1995,58 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+
+                    CPubKey pk(vchGenerator);
+                    CPubKey res(vchRes);
+                    if (vchGenerator.size() != CPubKey::COMPRESSED_SIZE || vchRes.size() != CPubKey::COMPRESSED_SIZE)
```
We can more directly check with `!pk.IsCompressed() || !res.IsCompressed()`.

Can you also add a test with an invalid public key (one that is a properly encoded compressed key, just not on the curve) and one with a bad encoding? I'm not sure if you need to do an `IsValid` or `IsFullyValid` check here, or if these are covered by `pk.TweakMulVerifiy`.

sanket1729 replied:
I was operating under the assumption that `scep` API should take care of all invalid inputs and the caller should not worry about these things. I will double-check.

sanket1729 replied:
It does do the check. Adding tests for checking this.

sanket1729 commented:
Pushed a diff
```diff
diff --git a/src/script/interpreter.cpp b/src/script/interpreter.cpp
index 843c8bb70..e7323b4cb 100644
--- a/src/script/interpreter.cpp
+++ b/src/script/interpreter.cpp
@@ -5,7 +5,7 @@
 
 #include <script/interpreter.h>
 
-#include <consensus/consensus.h>
+#include <consensus/validation.h>
 #include <crypto/ripemd160.h>
 #include <crypto/sha1.h>
 #include <crypto/sha256.h>
@@ -1807,11 +1807,11 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                     popstack(stack);
                     popstack(stack);
 
-                    auto inps = checker.GetTxvIn();
+                    auto inps = checker.GetTransaction()->vin;
                     auto spent_outputs = checker.GetPrecomputedTransactionData()->m_spent_outputs;
-                    if (idx < 0 || (unsigned int)idx >= inps->size() || (unsigned int)idx >= spent_outputs.size())
+                    if (idx < 0 || (unsigned int)idx >= inps.size() || (unsigned int)idx >= spent_outputs.size())
                         return set_error(serror, SCRIPT_ERR_INTROSPECT_INDEX_OUT_OF_BOUNDS);
-                    const CTxIn inp = inps->at(idx);
+                    const CTxIn inp = inps[idx];
                     const CTxOut spent_utxo = spent_outputs[idx];
 
                     switch (n)
@@ -1904,10 +1904,10 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                     popstack(stack);
                     popstack(stack);
 
-                    auto outs = checker.GetTxvOut();
-                    if (idx < 0 || (unsigned int)idx >= outs->size())
+                    auto outs = checker.GetTransaction()->vout;
+                    if (idx < 0 || (unsigned int)idx >= outs.size())
                         return set_error(serror, SCRIPT_ERR_INTROSPECT_INDEX_OUT_OF_BOUNDS);
-                    const CTxOut out = outs->at(idx);
+                    const CTxOut out = outs[idx];
 
                     switch (n)
                     {
@@ -1961,31 +1961,31 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                     {
                         case 0: // Version
                         {
-                            push4_le(stack, (uint32_t) checker.GetTxVersion());
+                            push4_le(stack, (uint32_t) checker.GetTransaction()->nVersion);
                             break;
                         }
                         case 1: // nLockTime
                         {
-                            push4_le(stack, checker.GetLockTime());
+                            push4_le(stack, checker.GetTransaction()->nLockTime);
                             break;
                         }
                         case 2: // Number of inputs
                         {
-                            if (checker.GetTxvIn()->size() > 0x7fffffff)
+                            if (checker.GetTransaction()->vin.size() > 0x7fffffff)
                                 return set_error(serror, SCRIPT_ERR_UNKNOWN_ERROR);
-                            push4_le(stack, checker.GetTxvIn()->size());
+                            push4_le(stack, checker.GetTransaction()->vin.size());
                             break;
                         }
                         case 3: // Number of outputs
                         {
                             // No need to bound checks for num_outputs. Elements consensus rules would break
                             // if outpoint more 2**30 is allowed.
-                            push4_le(stack, checker.GetTxvOut()->size());
+                            push4_le(stack, checker.GetTransaction()->vout.size());
                             break;
                         }
                         case 4: // Transaction size
                         {
-                            push4_le(stack, checker.GetTxWeight());
+                            push4_le(stack, GetTransactionWeight(*checker.GetTransaction()));
                             break;
                         }
                         default:
@@ -2005,7 +2005,7 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
 
                     CPubKey pk(vchGenerator);
                     CPubKey res(vchRes);
-                    if (!pk.IsCompressed() || !res.IsCompressed())
+                    if (vchGenerator.size() != CPubKey::COMPRESSED_SIZE || vchRes.size() != CPubKey::COMPRESSED_SIZE)
                         return set_error(serror, SCRIPT_ERR_PUBKEYTYPE);
                     execdata.m_validation_weight_left -= VALIDATION_WEIGHT_PER_SIGOP_PASSED;
                     if (execdata.m_validation_weight_left < 0) {
@@ -2105,9 +2105,9 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                             break;
                         case OP_DIV64:
                         {
-                            if (b == 0) { stack.push_back(vchFalse); break; }
                             int64_t r = a % b;
                             int64_t q = a / b;
+                            if (b == 0) { stack.push_back(vchFalse); break; }
                             if (r < 0 && b > 0)      { r += b; q-=1;} // ensures that 0<=r<|b|
                             else if (r < 0 && b < 0) { r -= b; q+=1;} // ensures that 0<=r<|b|
                             popstack(stack);
@@ -2919,34 +2919,9 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(const CScriptNum& nSeq
 }
 
 template <class T>
-unsigned int GenericTransactionSignatureChecker<T>::GetLockTime() const
-{
-    return txTo->nLockTime;
-}
-
-template <class T>
-unsigned int GenericTransactionSignatureChecker<T>::GetTxVersion() const
-{
-    return txTo->nVersion;
-}
-
-template <class T>
-const std::vector<CTxIn>* GenericTransactionSignatureChecker<T>::GetTxvIn() const
-{
-    return &(txTo->vin);
-}
-
-template <class T>
-const std::vector<CTxOut>* GenericTransactionSignatureChecker<T>::GetTxvOut() const
-{
-    return &(txTo->vout);
-}
-
-template <class T>
-unsigned int GenericTransactionSignatureChecker<T>::GetTxWeight() const
+const CTransaction* GenericTransactionSignatureChecker<T>::GetTransaction() const
 {
-    // line copied from GetTrasactionWeight() in src/consensus/validation.h
-    return ::GetSerializeSize(*txTo, PROTOCOL_VERSION | SERIALIZE_TRANSACTION_NO_WITNESS) * (WITNESS_SCALE_FACTOR - 1) + ::GetSerializeSize(*txTo, PROTOCOL_VERSION);
+    return (CTransaction*)(txTo);
 }
 
 template <class T>
diff --git a/src/script/interpreter.h b/src/script/interpreter.h
index 649577f08..13369a112 100644
--- a/src/script/interpreter.h
+++ b/src/script/interpreter.h
@@ -269,31 +269,11 @@ public:
          return false;
     }
 
-    virtual const std::vector<CTxIn>* GetTxvIn() const
+    virtual const CTransaction* GetTransaction() const
     {
         return nullptr;
     }
 
-    virtual const std::vector<CTxOut>* GetTxvOut() const
-    {
-        return nullptr;
-    }
-
-    virtual unsigned int GetLockTime() const
-    {
-        return 0;
-    }
-
-    virtual unsigned int GetTxVersion() const
-    {
-        return 0;
-    }
-
-    virtual unsigned int GetTxWeight() const
-    {
-        return 0;
-    }
-
     virtual const PrecomputedTransactionData* GetPrecomputedTransactionData() const
     {
         return nullptr;
@@ -327,11 +307,7 @@ public:
     bool CheckSchnorrSignature(Span<const unsigned char> sig, Span<const unsigned char> pubkey, SigVersion sigversion, const ScriptExecutionData& execdata, ScriptError* serror = nullptr) const override;
     bool CheckLockTime(const CScriptNum& nLockTime) const override;
     bool CheckSequence(const CScriptNum& nSequence) const override;
-    const std::vector<CTxIn>* GetTxvIn() const override;
-    const std::vector<CTxOut>* GetTxvOut() const override;
-    unsigned int GetLockTime() const override;
-    unsigned int GetTxVersion() const override;
-    unsigned int GetTxWeight() const override;
+    const CTransaction* GetTransaction() const override;
 
     const PrecomputedTransactionData* GetPrecomputedTransactionData() const override;
     unsigned int GetnIn() const override;
diff --git a/test/functional/feature_tapscript_opcodes.py b/test/functional/feature_tapscript_opcodes.py
index 4f0a843b9..113a6eaba 100755
--- a/test/functional/feature_tapscript_opcodes.py
+++ b/test/functional/feature_tapscript_opcodes.py
@@ -519,19 +519,13 @@ class TapHashPeginTest(BitcoinTestFramework):
         g = bytes.fromhex("0479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798")# y co-ordinate mutated
         self.tapscript_satisfy_test(CScript([OP_ECMULSCALAREXPVERIFY, OP_1]), inputs = [res, g, scalar], fail=True)
 
-        # test that other unknown keys(33 byte) fail correctly
+        # test that other random values fail correctly
         res = bytes.fromhex("032c0158d0f6df4881e99e65fbea21f27321d817f79ad39e08eaf4f16f1419bb0c")
         scalar = bytes.fromhex("e0f47c124f228b97bbdc0e4398aac9788869b9fbbc193d5323fdad9570609de6")
         g = bytes.fromhex("0979be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798")# Unkown keys
         self.tapscript_satisfy_test(CScript([OP_ECMULSCALAREXPVERIFY, OP_1]), inputs = [res, g, scalar], fail=True)
 
-        # test that other invalid out of curve points(33 byte) fail correctly
-        res = bytes.fromhex("032c0158d0f6df4881e99e65fbea21f27321d817f79ad39e08eaf4f16f1419bb0c")
-        scalar = bytes.fromhex("e0f47c124f228b97bbdc0e4398aac9788869b9fbbc193d5323fdad9570609de6")
-        g = bytes.fromhex("0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81799") #Invalid point not on curve
-        self.tapscript_satisfy_test(CScript([OP_ECMULSCALAREXPVERIFY, OP_1]), inputs = [res, g, scalar], fail=True)
-
-        # test that other 32 byte hashes fail correctly
+        # test that other random values fail correctly
         res = bytes.fromhex("032c0158d0f6df4881e99e65fbea21f27321d817f79ad39e08eaf4f16f1419bb0c")
         scalar = bytes.fromhex("e07c124f228b97bbdc0e4398aac9788869b9fbbc193d5323fdad9570609de6") # non 32 byte hash
         g = bytes.fromhex("0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798")

```

sanket1729 commented on src/crypto/sha256.cpp:
```diff
@@ -720,7 +720,7 @@ std::vector<unsigned char> CSHA256::Save() const {
 
     WriteLE64(&result[32], bytes << 3);
 
-    memcpy(&result[40], buf, buf_size);
```
Based on my reading here: https://stackoverflow.com/questions/29844298/is-it-legal-to-call-memcpy-with-zero-length-on-a-pointer-just-past-the-end-of-an 
I modified the implementation for a check for `memcpy` 0 bytes. 
Apparently `memcpy(src, object_end_ptr, 0)` is illegal and undefined behaviour. 

Added a separate commit for highlighting this part. Will squash later if needed.

roconnor-blockstream replied:
You can just say `if (buf_size)`, but this is also okay.

apoelstra commented:
utACK 6bf0da884afd2c1dbb9a0ba3d856abe4d7df7c19 

Running my local tests now, ni the morning I can probably do a full ACK.

cc @roconnor-blockstream this is ready for your review

apoelstra commented:
ack 6bf0da884afd2c1dbb9a0ba3d856abe4d7df7c19

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1587,29 +1689,533 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+
+                case OP_INSPECTCURRENTINPUT:
+                {
+                    // OP_INSPECTOUTPUT is available post tapscript
```
Bad comment.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1587,29 +1689,533 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    switch(opcode)
+                    {
+                        case OP_SUB64:
+                            b = -b;
```
Negating 0x8000...000 is UB.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1587,29 +1689,533 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                                stack.push_back(vchTrue);
+                            }
+                            break;
+                        case OP_DIV64:
```
`0x800000... / (-1)` appears to be UB.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1587,29 +1689,533 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    if (vchb.size() != 8 || vcha.size() != 8)
+                        return set_error(serror, SCRIPT_ERR_EXPECTED_8BYTES);
+
+                    int64_t b = ReadLE64(vchb.data());
```
ReadLE64 returns an `uint64_t`, but you are assigning it to an `int64_t`.  This conversion is "implementation defined".  I'm not sure how bad that is, but it doesn't sound good.  Does this really not yield a compilier warning?

sanket1729 replied:
Nope, there is no compiler warning

sanket1729 replied:
It looks like the functions for 
```
template<typename Stream> inline void Unserialize(Stream& s, int64_t& a) { a = ser_readdata64(s); }
``` in `serialize.h`
already do this implicit cast. So, we are safe in doing this over here?

roconnor-blockstream replied:
Well since you've written the cast function already we might as well use it.  I'd like to see how this shakes out in Bitcoin Core land, but ya, a direct cast is probably also acceptable.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1587,29 +1689,533 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                        case OP_GREATERTHAN64:         popstack(stack); popstack(stack); stack.push_back( (a >  b) ? vchTrue : vchFalse ); break;
+                        case OP_GREATERTHANOREQUAL64:  popstack(stack); popstack(stack); stack.push_back( (a >= b) ? vchTrue : vchFalse ); break;
+                        case OP_EQUAL64:               popstack(stack); popstack(stack); stack.push_back( (a == b) ? vchTrue : vchFalse ); break; //same as op_equal
+                        case OP_AND64:                 popstack(stack); popstack(stack); push8_le(stack, a & b); break;
```
I don't know.  Maybe for these bitwise operations we should cast to unsigned just to be safe.  The cast should be a nop in practice.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1587,29 +1689,533 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    }
+                }
+                break;
+                case OP_NOT64:
```
Why is NOT64 bundled with the shifts?

sanket1729 replied:
These all operate on a single operand. These were grouped together to avoid some repeat checks on the single operand.

roconnor-blockstream replied:
But the shifts are binary operations.  Oh the shift amounts are ScriptNums.  Got it.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -60,6 +61,67 @@ static inline void popstack(std::vector<valtype>& stack)
+    stack.push_back(vch);
+}
+
+static inline void push8_le(std::vector<valtype>& stack, int64_t v)
```
Probably should be `uint64_t v` parameter.

sanket1729 replied:
Sure, there will be implicit cast from `signed` to `unsigned` somewhere. Don't really have a preference whether it's in a function body or parameter.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -60,6 +61,67 @@ static inline void popstack(std::vector<valtype>& stack)
+{
+    uint64_t int64_min = (uint64_t)std::numeric_limits<int64_t>::min();
+    if (v >= int64_min)
+        return static_cast<int64_t>(v - int64_min) + int64_min;
```
return static_cast<int64_t>(v - int64_min) + std::numeric_limits<int64_t>::min();

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1587,29 +1695,538 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                        case OP_INSPECTOUTPUTSCRIPTPUBKEY:
+                        {
+                            valtype vchScriptPubKey;
+                            vchScriptPubKey.insert(vchScriptPubKey.begin(), out.scriptPubKey.begin(), out.scriptPubKey.end());
```
What happens if/when the max item size is exceeded?

sanket1729 replied:
Good catch, this should be checked here. I was only considering taproot outputs which are 34 bytes. Nut this introspection allows inspecting other outputs too which could be more than 520. And potentially this can be a DoS vector.

roconnor-blockstream replied:
We could return the hash of the pubkey, though that requires some caching.
Maybe we should return a parsed pubkey, with the segwit version pushed, and the segwit program pushed, and -1 pushed if it isn't native segwit.

sanket1729 replied:
I like this idea. We just push the version and program separately. They are also constrained to 40 bytes by BIP141.

roconnor-blockstream replied:
Since people may want to check their OP_RETURN data outputs too, we probably want both.

I don't know, maybe for non-segwit outputs we should push a hash of the pubkey, and for segwit output we push the parsed version?  Maybe it is okay to just hash in place to start with and optimize with a cache later?

sanket1729 replied:
In the non-segwit case, How about we fail if `spk > MAX_SCRIPT_ELEMENT_SIZE` and push the script pubkey as it is?

`OP_RETURN`s are restricted by `MAX_OP_RETURN_RELAY = 83` by policy rules, so I don't think we are restricting things severely by doing this.

apoelstra commented on src/script/script_error.cpp:
```diff
@@ -119,6 +119,10 @@ std::string ScriptErrorString(const ScriptError serror)
         case SCRIPT_ERR_INTROSPECT_INDEX_OUT_OF_BOUNDS:
             return "Introspection index out of bounds";
+        case SCRIPT_ERR_EXPECTED_8BYTES:
+            return "Arethmetic opcodes expect 8 bytes operands";
```
spelling `arithmetic` (also in the commit message)

apoelstra commented:
utACK 2064492ceb79c9962436a48b86f5fa76f5d607d1

This looks great

apoelstra commented:
ack 2064492ceb79c9962436a48b86f5fa76f5d607d1

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1587,29 +1710,510 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+
+                            // Push the outpoint flag
+                            valtype vchOutpointFlag(1);
+                            vchOutpointFlag[0] = (unsigned char) ((!inp.assetIssuance.IsNull() << 7) + (inp.m_is_pegin << 6));
```
Didn't we define a function for this computation?

sanket1729 replied:
That is not merged yet :( #1023

apoelstra commented:
reminder to add a commit to change the tapleaf version

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -459,6 +459,47 @@ static bool EvalChecksigPreTapscript(const valtype& vchSig, const valtype& vchPu
+        return set_error(serror, SCRIPT_ERR_PUBKEYTYPE);
+    } else if (pubkey.size() == 32) {
+        if (success && !pubkey.VerifySchnorr(msg, sig)) {
+            return false; // serror is set
```
No `serror` has not been set.

I think test coverage must be lacking here.

sanket1729 replied:
yes, it is. I have only asserted the failure of the test in the test suite and not the actual error code for it. I will update the tests so that they assert the real reason why they fail.

roconnor-blockstream replied:
Ah right. This error is less severe than I had thought.

sanket1729 replied:
In any case, I am updating the test suite to check for actual return error codes.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1587,29 +1710,510 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    if (stack.size() < 2)
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+
+                    valtype& vchb = stacktop(-1);
```
Please reverse the order of these `stacktop` items.  I know we are not always consistent, but let's try to keep `stack(-1)` at the end.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1953,6 +2011,58 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    const XOnlyPubKey tweakedKey{uint256(std::vector<unsigned char>(vchTweakedKey.begin() + 1, vchTweakedKey.begin() + CPubKey::COMPRESSED_SIZE))};
+                    const uint256 tweak(vchTweak);
+                    const XOnlyPubKey internalKey{uint256(vchInternalKey)};
+                    if (!tweakedKey.CheckPayToContract(internalKey, tweak, vchTweakedKey[0] & 1))
```
AFAICT there is not check that vchTweakedKey[0] is either 0x02 or 0x03.

More testing?

sanket1729 replied:
Great catch! This is indeed a malleability vector.

apoelstra replied:
This is checked above in the `IsFullyValid` check on `tweakedKey`

sanket1729 replied:
That was added after Russell pointed this out.

sanket1729 commented:
Updated the spec in #1019 and updated the tests to check the exact error message instead of just asserting failure. 
Ready for another round of review @apoelstra @roconnor-blockstream

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -467,6 +467,47 @@ static bool EvalChecksigPreTapscript(const valtype& vchSig, const valtype& vchPu
+    if (pubkey.size() == 0) {
+        return set_error(serror, SCRIPT_ERR_PUBKEYTYPE);
+    } else if (pubkey.size() == 32) {
+        if (success && !pubkey.VerifySchnorr(msg, sig)) {
```
`VerifySchnorr` has a precondition that `sig` is 64-bytes, but I don't see a check for this precondition.

More testing?

sanket1729 replied:
Indeed, I tested empty sigs, mutated 64-byte sigs, but did not check 64-byte sigs :(

apoelstra replied:
:/ surprised the fuzzer didn't hit this given that it's an actual assertion failure and not even guarded by any cryptography

apoelstra commented on src/script/interpreter.h:
```diff
@@ -173,6 +173,8 @@ struct PrecomputedTransactionData
     uint256 m_issuance_rangeproofs_single_hash;
+    //! transaction outputs scriptpubkey sha single hash
+    std::vector<uint256> m_output_spk_single_hashes;
     //! Whether the 10 fields above are initialized.
```
nit: this comment should now say "11 fields"

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1947,9 +2085,10 @@ std::vector<uint256> GetSpentScriptPubKeysSHA256(const std::vector<CTxOut>& outp
+        // Normal serialization using the << operater would also serialize the lenght, therefore we directly write using CSHA256
+        uint256 spent_spk_single_hash;
+        CSHA256().Write(txout.scriptPubKey.data(), txout.scriptPubKey.size()).Finalize(spent_spk_single_hash.begin());
+        spent_spk_single_hashes.push_back(spent_spk_single_hash);
```
For what it's worth, I think it was fine to serialize the length. Also spelling `length` in the comment.

apoelstra commented on test/functional/test_framework/script.py:
```diff
@@ -320,7 +336,7 @@ def __new__(cls, n):
     OP_2DUP: 'OP_2DUP',
     OP_3DUP: 'OP_3DUP',
-    OP_2OVER: 'OP_2OVER',
+OP_2OVER: 'OP_2OVER',
```
this looks like an accidental diff

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1672,29 +1721,41 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                            return set_error(serror, SCRIPT_ERR_CHECKSIGVERIFY);
+                    } else {
+                        // New BIP 340 semantics for CHECKSIGFROMSTACK
+                        const XOnlyPubKey pubkey{uint256(vchPubKey)};
```
`uint256` constructor asserts that its input is 32 bytes but there is no constraint on `vchPubKey`. (Should also add a test for this)

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -467,6 +467,51 @@ static bool EvalChecksigPreTapscript(const valtype& vchSig, const valtype& vchPu
     return true;
 }
 
+static bool EvalTapScriptCheckSigFromStack(const valtype& sig, const valtype& vchPubKey, ScriptExecutionData& execdata, unsigned int flags, const Span<const unsigned char> msg, SigVersion sigversion, ScriptError* serror, bool& success)
```
`const valtype& msg`.  In particular it should be passed by reference?

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -467,6 +467,51 @@ static bool EvalChecksigPreTapscript(const valtype& vchSig, const valtype& vchPu
+        if (success) {
+            if (sig.size() != 64)
+                return set_error(serror, SCRIPT_ERR_SCHNORR_SIG_SIZE);
+            const XOnlyPubKey pubkey{uint256(vchPubKey)};
```
No need to pass through `uint256` here AFAICT.

sanket1729 replied:
Changed here and other places

apoelstra commented:
ack 3ad65cea1020c692840675bf8e3f458f34aeb69d

Will run the new fuzztest for several more hours.

apoelstra commented:
utACK 385a6bb4d2929106f201f0114abb0f11dda6708b

Confirmed it is a rename-only change from 3ad65cea1020c692840675bf8e3f458f34aeb69d which I fully acked.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1962,6 +2023,60 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    if (execdata.m_validation_weight_left < 0) {
+                        return set_error(serror, SCRIPT_ERR_TAPSCRIPT_VALIDATION_WEIGHT);
+                    }
+                    const XOnlyPubKey tweakedXOnlyKey{std::vector<unsigned char>(vchTweakedKey.begin() + 1, vchTweakedKey.begin() + CPubKey::COMPRESSED_SIZE)};
```
Seems better to use a span here, but what do I know.  At least it isn't as bad as https://github.com/bitcoin/bitcoin/blob/b6a8e68b4e11be3aa5ceff8c483594f8e576faaf/src/script/interpreter.cpp#L1877.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1962,6 +2023,60 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    valtype& vchInternalKey = stacktop(-1);
+
+                    CPubKey tweakedKey(vchTweakedKey);
+                    if (!tweakedKey.IsFullyValid() || !tweakedKey.IsCompressed() || vchInternalKey.size() != 32 || vchTweak.size() != 32)
```
Checking FullyValid is an unnecessary expense.  I believe IsCompressed is good enough.

sanket1729 replied:
I did consider this, but there is still some utility in that this returns a different error type.

sanket1729 replied:
I can also just substitute it for the simple check for the first byte 0x02

sanket1729 replied:
This is more efficient, part of me was over checking things because of the aborts :)

roconnor-blockstream replied:
IsCompressed is fine.  It implies 0x02 and 0x03.  My concern with FullyValid is that it involves the unprecedented and moderately expensive crypto operation of sqrt.

sanket1729 replied:
I directly checked the first byte instead of allocating a new `CPubKey` structure which was not required.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1962,6 +2023,60 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+
+                    CPubKey pk(vchGenerator);
+                    CPubKey res(vchRes);
+                    if (!pk.IsFullyValid() || !res.IsFullyValid() || !pk.IsCompressed() || !res.IsCompressed())
```
Same here, Checking FullyValid is an unnecessary expense.  IsCompressed should be good enough.

roconnor-blockstream commented on src/pubkey.cpp:
```diff
@@ -188,6 +188,17 @@ bool XOnlyPubKey::CheckPayToContract(const XOnlyPubKey& base, const uint256& has
     return secp256k1_xonly_pubkey_tweak_add_check(secp256k1_context_verify, m_keydata.begin(), parity, &base_point, hash.begin());
 }
 
+bool CPubKey::TweakMulVerify(const CPubKey& res, const uint256& tweak) const
```
I moderately prefer this method is deleted and its operation is inlined.  TweakMulVerify as written doesn't really handle non-compressed keys, but as a method of CPubKey it would need to.

sanket1729 replied:
Agreed

sanket1729 replied:
On second thought,

- There is also derive method on `CPubKey` that does not make sense for Uncompressed keys. I can add an assert to ensure it. 
- The current code architecture makes it cleaner to use all libsecp methods inside the pubkey module. This code would be the only code in `interpreter.cpp` that would require context objects, and directly accesses the libsecp methods. 
- Agreed with naming. I can interchange the position of `res` and `this` in the method.

roconnor-blockstream replied:
Fairpoint about the encapsulation of secp256k1.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1949,6 +1962,160 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                            }
+                        break;
+                        case OP_SUB64:
+                            if ((a > 0 && b <= std::numeric_limits<int64_t>::min() + a) ||
```
This expression is pretty confusing.

    (b > 0 && a < std::numeric_limits<int64_t>::min() + b) || (b < 0 && a > std::numeric_limits<int64_t>::max() + b)

sanket1729 replied:
Yes, this is much cleaner. I somehow had my mind fixated on branching on a > 0 and a < 0.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1949,6 +1962,160 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+
+                    int num = CScriptNum(stacktop(-1), fRequireMinimal).getint();
+                    popstack(stack);
+                    push8_le(stack, num);
```
I think this line is correct.  I'd feel much more comfortable if num had an `int64_t` type.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1941,6 +1941,33 @@ uint256 GetSpentScriptsSHA256(const std::vector<CTxOut>& outputs_spent)
+
+/** Compute the vector where each element is SHA256 of output scriptPubKey of a tx. */
+template <class T>
+std::vector<uint256> GetOutputScriptPubKeysSHA256(const T& txTo)
```
Maybe take `const std::vector<CTxOut>& vout` parameter to avoid the template?

roconnor-blockstream replied:
Maybe it doesn't matter.

sanket1729 replied:
This is consistent with `GetOutputsSHA256` is implemented.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -61,6 +61,61 @@ static inline void popstack(std::vector<valtype>& stack)
+    }
+}
+
+/** Compute the outpoint flag(u8) for a given txin **/
```
Needs a rebase?

roconnor-blockstream replied:
Oh nevermind. I now see this is moved code.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1947,9 +2085,10 @@ std::vector<uint256> GetSpentScriptPubKeysSHA256(const std::vector<CTxOut>& outp
-        CHashWriter ss(SER_GETHASH, 0);
-        ss << txout.scriptPubKey;
-        spent_spk_single_hashes.push_back(ss.GetSHA256());
+        // Normal serialization using the << operater would also serialize the length, therefore we directly write using CSHA256
```
Shouldn't these be changes be made in the original commit?

roconnor-blockstream commented on src/script/script_error.h:
```diff
@@ -90,7 +90,8 @@ typedef enum ScriptError_t
+    SCRIPT_ERR_INTROSPECT_INDEX_OUT_OF_BOUNDS
 } ScriptError;
 
 #define SCRIPT_ERR_LAST SCRIPT_ERR_ERROR_COUNT
```
I'm glad SCRIPT_ERR_LAST isn't used anywhere.

sanket1729 commented:
I think have addressed all comments by either fixing them or responding as to why it is better to leave them as is.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -15,6 +15,9 @@
 typedef std::vector<unsigned char> valtype;
 
+static_assert(std::numeric_limits<int64_t>::max() == (int64_t)0x7FFFFFFFFFFFFFFF);
+static_assert(std::numeric_limits<int64_t>::min() == (int64_t)-0x8000000000000000);
```
I'd rather have

    static_cast<uint64_t>(std::numeric_limits<int64_t>::max())==UINT64_C(0x7FFF'FFFF'FFFF'FFFF)
    static_cast<uint64_t>(std::numeric_limits<int64_t>::min())==UINT64_C(0x8000'0000'0000'0000)

sanket1729 replied:
macros are incompatible with the ' separation. Changed to 
```
static_assert(static_cast<uint64_t>(std::numeric_limits<int64_t>::max()) == UINT64_C(0x7FFFFFFFFFFFFFFF));
static_assert(static_cast<uint64_t>(std::numeric_limits<int64_t>::min()) == UINT64_C(0x8000000000000000));
```

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1743,7 +1759,7 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
                 case OP_INSPECTINPUTSEQUENCE:
                 case OP_INSPECTINPUTISSUANCE:
                 {
-                    // OP_INSPECTINPUT is available post tapscript
```
Can we move all these changes, including the push4_le, back into the previous commit?

sanket1729 replied:
Done. Squashed introspection-related commits into one.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1827,6 +1837,118 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    // OP_PUSHCURRENTINPUTINDEX is available post tapscript
+                    if (sigversion == SigVersion::BASE || sigversion == SigVersion::WITNESS_V0) return set_error(serror, SCRIPT_ERR_BAD_OPCODE);
+
+                    if (checker.GetnIn() > 0x7fffffff)
```
We can probably make this an assert.  Having more that MAX_SIZE (2^25) inputs is illegal.

roconnor-blockstream replied:
Apparently In BaseSignatureChecker this is set to 0xFFFFFFFF in exotic contexts, so I guess we cannot assume anything.  I guess we can keep it as is.

roconnor-blockstream replied:
Maybe add a comment.

roconnor-blockstream replied:
I'm not a big fan of the use of this sentinel values.  the 0x7ffff is just too arbitrary.  There are a variety of things that I think would be better.

1. Testing against 0xffffff and assigning it to a static constant in the BaseSignatureChecker class.
2. Comparing with MAX_SIZE instead of 0x7ffffff here.
3. Having GetnIn return a const pointer to an unsigned int and return nullptr in BaseSignatureChecker.
4. Having GetnInt return an Optional integer.

roconnor-blockstream replied:
5. have GetnIn return an int or a int32_t and check if the value is negative.

I think this one is my favourite.

sanket1729 replied:
I think checking with MAX_SIZE makes some sense. It would still require casting `unsigned int nIn` to some `int` type.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1752,205 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+
+                    if (checker.GetnIn() > 0x7fffffff)
+                        return set_error(serror, SCRIPT_ERR_UNKNOWN_ERROR);
+                    CScriptNum currIdx = CScriptNum((int64_t) checker.GetnIn());
```
I believe static_casts are preferred in C++.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1752,205 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                                stack.push_back(vchAssetEntropy);
+                                // Finally push blinding nonce
+                                // By pushing the this order, we make sure that the stack top is empty
+                                // iff there is no issuance. Note that nInflationKeys can be null and can push false
```
Either one of inflation keys or the asset amount could be Null.

sanket1729 replied:
Updated, it was my misunderstanding that these outputs should be non-zero.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1752,205 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                        }
+                        case OP_INSPECTOUTPUTSCRIPTPUBKEY:
+                        {
+                            valtype vchScriptPubKeySha256;
```
This and everywhere else should be constructor initialized right?

sanket1729 replied:
Created vchStackItem with a new constructor `vchItem(begin(), end())` which is much cleaner to read.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1752,205 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    auto outs = checker.GetTxvOut();
+                    if (idx < 0 || (unsigned int)idx >= outs->size())
+                        return set_error(serror, SCRIPT_ERR_INTROSPECT_INDEX_OUT_OF_BOUNDS);
+                    const CTxOut out = outs->at(idx);
```
Should be a reference?

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1752,205 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    auto spent_outputs = cache->m_spent_outputs;
+                    if (idx < 0 || (unsigned int)idx >= inps->size() || (unsigned int)idx >= spent_outputs.size())
+                        return set_error(serror, SCRIPT_ERR_INTROSPECT_INDEX_OUT_OF_BOUNDS);
+                    const CTxIn inp = inps->at(idx);
```
Should be references?

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -61,6 +61,77 @@ static inline void popstack(std::vector<valtype>& stack)
+{
+    valtype vch;
+    auto v_le = htole32(v);
+    vch.insert(vch.begin(), (unsigned char*)&v_le, (unsigned char*)&v_le + 4);
```
reinterpret casts?

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1752,205 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    int idx = CScriptNum(stacktop(-1), fRequireMinimal).getint();
+                    popstack(stack);
+
+                    auto inps = checker.GetTxvIn();
```
These pointers could be null in exotic SignatureChecker contexts.  I don't think we can assume they are non-null.

sanket1729 replied:
Returning error `SCRIPT_ERR_INTROSPECT_CONTEXT_UNAVAILABLE` to reflect this

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1752,205 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    popstack(stack);
+
+                    const PrecomputedTransactionData *cache = checker.GetPrecomputedTransactionData();
+                    auto outs = checker.GetTxvOut();
```
This pointer could be null in exotic SignatureChecker contexts.  I don't think we can assume it is non-null.

apoelstra replied:
There are no exotic SignatureChecker contexts except in testing envionments.

apoelstra replied:
Actually, in Elements there is, because we have a script interpreter in blockheaders :).

roconnor-blockstream replied:
https://github.com/bitcoin/bips/blob/master/bip-0322.mediawiki

roconnor-blockstream replied:
Regardless you cannot have a post-condition for BaseSignatureChecker be that the pointer is non-null and then implement it with "return nullptr".  Either assert false or make it pure virtual, or deal with the fact that the post-condition is that the pointer can be null.

Requiring a whole program analysis to understand the correctness of this is not acceptable.

roconnor-blockstream replied:
Given that we have to live with BaseSignatureChecker's inheritance hierarchy mess for the time being, the way I'd like to see this improved is to have

1. have these pointer returning methods return references instead of pointers.
2. assert false for most or all of these default methods (instead of returning null pointers).

roconnor-blockstream replied:
The alternative is to explicitly handle these null pointers in EvalScript, by calling seterror and returning false.

sanket1729 replied:
This is what is currently implemted. I think the correct solution is to return a type `Option<reference>`, but I suspect that will leak throughout the codebase.

roconnor-blockstream replied:
Okay after looking into (legacy) `DataFromTransaction`, I now believe what we have written is, more or less, the best we can do.

However we need some sort of TODO item here to assert false depending on the MissingDataBehavior setting after it is introduced in https://github.com/bitcoin/bitcoin/pull/21330.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1613,6 +1613,73 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+
+                    valtype& vch = stacktop(-2);
+                    if (!ctx.SafeWrite(vch.data(), vch.size()))
+                        return set_error(serror, SCRIPT_ERR_UNKNOWN_ERROR);
```
Maybe we could find/add a better error message for all these `SafeWrite` failures?

sanket1729 replied:
Added `SCRIPT_ERR_SHA2_CONTEXT_WRITE`

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1613,6 +1613,73 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    if (stack.size() < 2)
+                        return set_error(serror, SCRIPT_ERR_INVALID_STACK_OPERATION);
+
+                    valtype& vchCtx = stacktop(-1);
```
I don't know why you reversed the order of the stack inputs here, but you didn't update the comments to match this change.

    // (sha256_ctx in -- hash)

Don't we want to take a context, then introspect some data or whatnot, and the merge the new data into the hash context?  I think that is going to be more common than taking witness data and adding it to a SHA56 context.

sanket1729 replied:
> But why did you reverse the order of the inputs to begin with? Don't we want to take a context, then introspect some data or whatnot, and merge the new data into the hash context? I think that is going to be more common than taking witness data and adding it to a SHA56 context.

Idk why I was thinking that taking in input from witnesses would be common, but I think you are correct. Updated to reverse the order.

apoelstra commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1752,205 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                case OP_INSPECTNUMOUTPUTS:
+                case OP_TXWEIGHT:
+                {
+                    // Transaction intropsection is available post tapscript
```
typo `introspection`

sanket1729 replied:
fixed

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1613,6 +1613,73 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+
+                case OP_SHA256FINALIZE: // (sha256_ctx in -- hash)
+                {
+                    // OP_SHA256Finalize is only available in Tapscript
```
Nit: Capitalize Finalize.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1613,6 +1613,73 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    CSHA256 ctx;
+                    valtype& vch = stacktop(-1);
+                    if (!ctx.SafeWrite(vch.data(), vch.size()))
+                        return set_error(serror, SCRIPT_ERR_UNKNOWN_ERROR);
```
SCRIPT_ERR_SHA2_CONTEXT_WRITE.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1613,6 +1613,73 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    if (!ctx.SafeWrite(vch.data(), vch.size()))
+                        return set_error(serror, SCRIPT_ERR_SHA2_CONTEXT_WRITE);
+
+                    valtype vchHash(32);
```
Nit: Should be vchHash(CSHA256::OUTPUT_SIZE).

roconnor-blockstream commented on src/script/script.cpp:
```diff
@@ -416,5 +420,6 @@ bool IsOpSuccess(const opcodetype& opcode)
            (opcode >= 141 && opcode <= 142) || (opcode >= 149 && opcode <= 151) ||
            // ELEMENTS: Exclude OP_DETERMINISTICRANDOM, OP_CHECKSIGFROMSTACK(VERIFY), OP_SUBSTRLAZY
-           (opcode >= 187 && opcode <= 191) || (opcode >= 196 && opcode <= 254);
+           // ELEMENTS: Tapscript extension till
```
extension till?

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -2398,6 +2399,49 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(const CScriptNum& nSeq
+}
+
+template <class T>
+unsigned int GenericTransactionSignatureChecker<T>::GetTxWeight() const
```
Return a (u)int64_t (similar to GetTransactionWeight)  instead of a `unsigned int`.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -2398,6 +2399,49 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(const CScriptNum& nSeq
+template <class T>
+unsigned int GenericTransactionSignatureChecker<T>::GetTxWeight() const
+{
+    // line copied from GetTrasactionWeight() in src/consensus/validation.h
```
s/GetTrasactionWeight/GetTransactionWeight/

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -2398,6 +2399,49 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(const CScriptNum& nSeq
+}
+
+template <class T>
+unsigned int GenericTransactionSignatureChecker<T>::GetTxVersion() const
```
Return an (u)int32_t.

sanket1729 replied:
This now returns int32_t, when is later cast to uint32_t before writing on the stack.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -2398,6 +2399,49 @@ bool GenericTransactionSignatureChecker<T>::CheckSequence(const CScriptNum& nSeq
 }
 
+template <class T>
+unsigned int GenericTransactionSignatureChecker<T>::GetLockTime() const
```
Return a uint32_t.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -2012,6 +2041,8 @@ void PrecomputedTransactionData::Init(const T& txTo, std::vector<CTxOut>&& spent
         m_issuance_rangeproofs_single_hash = GetIssuanceRangeproofsSHA256(txTo);
         m_output_witnesses_single_hash = GetOutputWitnessesSHA256(txTo);
         m_spent_scripts_single_hash = GetSpentScriptsSHA256(m_spent_outputs);
+        m_spent_output_spk_single_hashes = GetSpentScriptPubKeysSHA256(m_spent_outputs);
```
Shouldn't there be a std::move here or something?  Not that I know anything about move semantics.

apoelstra replied:
It should be fine. Any compiler since 2000 or so is smart enough to do "copy elision" when assigning the return value of a function like this.

sanket1729 replied:
I was assuming that there would be some return value optimization that would apply the move semantics here. 
https://stackoverflow.com/questions/15704565/efficient-way-to-return-a-stdvector-in-c

sanket1729 replied:
Updated the PR to use std::move

roconnor-blockstream replied:
Probably should turn it back.  It sounds like std:move for return values is somewhere between neutral and bad.

sanket1729 replied:
Done

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -2455,9 +2728,9 @@ const std::vector<CTxOut>* GenericTransactionSignatureChecker<T>::GetTxvOut() co
 
 template <class T>
-unsigned int GenericTransactionSignatureChecker<T>::GetTxWeight() const
+uint64_t GenericTransactionSignatureChecker<T>::GetTxWeight() const
```
These fixes are in the wrong commit.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -61,6 +64,19 @@ static inline void popstack(std::vector<valtype>& stack)
 
+static inline int64_t cast_signed64(uint64_t v)
+{
+    uint64_t int64_min = (uint64_t)std::numeric_limits<int64_t>::min();
```
This should be a `static_cast`.

roconnor-blockstream commented on src/pubkey.cpp:
```diff
@@ -188,6 +188,18 @@ bool XOnlyPubKey::CheckPayToContract(const XOnlyPubKey& base, const uint256& has
+    unsigned char out_pk[CPubKey::COMPRESSED_SIZE];
+    size_t out_len = CPubKey::COMPRESSED_SIZE;
+    if (!secp256k1_ec_pubkey_serialize(secp256k1_context_verify, out_pk, &out_len, &pk, SECP256K1_EC_COMPRESSED)) return false;
+    return *this == CPubKey(out_pk, out_pk + CPubKey::COMPRESSED_SIZE);
```
Perhaps `assert(CPubKey::COMPRESSED_SIZE == out_len)`, or use `out_pk + out_len` instead of `out_pk + CPubKey::COMPRESSED_SIZE`.

roconnor-blockstream commented on src/pubkey.cpp:
```diff
@@ -188,6 +188,18 @@ bool XOnlyPubKey::CheckPayToContract(const XOnlyPubKey& base, const uint256& has
 
+bool CPubKey::TweakMulVerify(const CPubKey& untweaked, const uint256& tweak) const
+{
+    assert(this->IsCompressed() && untweaked.IsCompressed());
```
I'm not sure that `untweaked.IsCompressed()` is a precondition.

apoelstra replied:
Agreed, no need to assert this, but for future reviewers, be aware that the sole use of this function does actually enforce that `untweaked` is compressed.

apoelstra commented:
ACK 56067a2983d8459849708ded3428944d2a23d93a

roconnor-blockstream commented on src/script/script.cpp:
```diff
@@ -151,6 +151,23 @@ std::string GetOpName(opcodetype opcode)
     case OP_SHA256INITIALIZE       : return "OP_SHA256INITIALIZE";
     case OP_SHA256UPDATE           : return "OP_SHA256UPDATE";
     case OP_SHA256FINALIZE         : return "OP_SHA256FINALIZE";
+    case OP_INSPECTINPUTOUTPOINT   : return "OP_INSPECTINPUTOUTPOINT";
```
I think these are in the wrong commit?

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -2457,7 +2730,7 @@ const std::vector<CTxOut>* GenericTransactionSignatureChecker<T>::GetTxvOut() co
 uint64_t GenericTransactionSignatureChecker<T>::GetTxWeight() const
 {
-    // line copied from GetTrasactionWeight() in src/consensus/validation.h
+    // line copied from GetTransactionWeight() in src/consensus/validation.h
```
This fixup is in the wrong commit.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1941,6 +1941,35 @@ uint256 GetSpentScriptsSHA256(const std::vector<CTxOut>& outputs_spent)
+    for (const auto& txout : outputs_spent) {
+        // Normal serialization using the << operater would also serialize the length, therefore we directly write using CSHA256
+        uint256 spent_spk_single_hash;
+        CSHA256().Write(txout.scriptPubKey.data(), txout.scriptPubKey.size()).Finalize(spent_spk_single_hash.begin());
```
Nit:  I think using `spent_spk_single_hash.data()` would be more appropriate, but what you have is probably still acceptable.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1941,6 +1941,35 @@ uint256 GetSpentScriptsSHA256(const std::vector<CTxOut>& outputs_spent)
+    for (const auto& txout : txTo.vout) {
+        // Normal serialization using the << operater would also serialize the length, therefore we directly write using CSHA256
+        uint256 out_spk_single_hash;
+        CSHA256().Write(txout.scriptPubKey.data(), txout.scriptPubKey.size()).Finalize(out_spk_single_hash.begin());
```
Nit:  I think using `out_spk_single_hash.data()` would be more appropriate, but what you have is probably still acceptable.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -61,6 +61,75 @@ static inline void popstack(std::vector<valtype>& stack)
+{
+    valtype vchinpValue;
+    if (value.IsNull()) {
+        vchinpValue.insert(vchinpValue.begin(), 8, 0);
```
pieter suggested that `assign` is more appropriate than `insert` in these cases where you cannot do constructor initialization.  But otherwise, where possible, you should replace your use of `insert` with constructor initialization.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1750,218 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    // TODO: Handle accoding to MissingDataBehavior
+                    if (!inps || !cache)
+                        return set_error(serror, SCRIPT_ERR_INTROSPECT_CONTEXT_UNAVAILABLE);
+                    auto spent_outputs = cache->m_spent_outputs;
```
Shouldn't this be a reference rather than copying the vector?

roconnor-blockstream replied:
Or is auto a reference?

sanket1729 replied:
This is not an auto reference. Verified that c++ actually copies the vector when we use `auto`

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1750,218 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    if (!inps || !cache)
+                        return set_error(serror, SCRIPT_ERR_INTROSPECT_CONTEXT_UNAVAILABLE);
+                    auto spent_outputs = cache->m_spent_outputs;
+                    if (idx < 0 || static_cast<unsigned int>(idx) >= inps->size() || static_cast<unsigned int>(idx) >= spent_outputs.size())
```
spent_outputs.size should be equal to inputs size.  Maybe you should be asserting that.

roconnor-blockstream replied:
Actually we also need an return false if cache.m_spent_outputs_ready is false, with a TODO to make it assert based on `mdb` value.

See for example, https://github.com/bitcoin/bitcoin/blob/eb09c26724e3f714b613788fc506f2ff3a208d2c/src/script/interpreter.cpp#L1523-L1525

sanket1729 replied:
Adding an assert after checking `m_spent_outputs`

roconnor-blockstream replied:
Why not add an assert before the `if` statement and remove one of the clauses?

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1750,218 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                            push4_le(stack, inp.prevout.n);
+
+                            // Push the outpoint flag
+                            valtype vchOutpointFlag(1);
```
`valtype vchOutpointFlag(1,GetOutpointFlag(inp));`
or
`stack.push_back(valtype(1,GetOutpointFlag(inp));`

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1750,218 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                    {
+                        case OP_INSPECTINPUTOUTPOINT:
+                        {
+                            valtype vchPrevTxid = valtype(inp.prevout.hash.begin(),inp.prevout.hash.end());
```
`valtype vchPrevTxid(inp.prevout.hash.begin(),inp.prevout.hash.end());`

or

`stack.push_back(valtype(inp.prevout.hash.begin(),inp.prevout.hash.end());`

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1750,218 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                        }
+                        case OP_INSPECTNUMINPUTS:
+                        {
+                            uint64_t num_ins = checker.GetTxvIn()->size();
```
You have to check to see if `checker.GetTxvIn()` is null and return false in that case, with a TODO to assert failure based on the `mdb` value.

roconnor-blockstream replied:
This type should be valtype::size_type I believe.

sanket1729 replied:
Is it a bug to do an implicit cast from `size_t` to `uint64_t`

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1750,218 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                        {
+                            // No need to bound checks for num_outputs. Elements consensus rules would break
+                            // if outpoint more 2**30 is allowed.
+                            uint64_t num_outs = checker.GetTxvOut()->size();
```
You have to check to see if `checker.GetTxvOut()` is null and return false in that case, with a TODO to assert failure based on the `mdb` value.

roconnor-blockstream replied:
This type should be valtype::size_type I believe.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -1681,6 +1750,218 @@ bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript&
+                            if (out.nNonce.IsNull()) {
+                                stack.push_back(vchFalse);
+                            } else {
+                                vchOutNonce.insert(vchOutNonce.begin(), out.nNonce.vchCommitment.begin(), out.nNonce.vchCommitment.begin() + 33);
```
vchOutNonce can be declared in here, and the insert turned into a constructor.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -61,6 +61,75 @@ static inline void popstack(std::vector<valtype>& stack)
+static inline void push4_le(std::vector<valtype>& stack, uint32_t v)
+{
+    auto v_le = htole32(v);
+    valtype vch(reinterpret_cast<unsigned char*>(&v_le), reinterpret_cast<unsigned char*>(&v_le) + 4);
```
emplace_back.

sanket1729 replied:
For my understanding, 

The difference between
```
valtype vch(args);
stack.push_back(std::move(vch));
```
and
```
stack.emplace_back(args);
```
is just a cheap move operation.

sanket1729 replied:
The `emplace_back` seems unreadable to my eyes, so I tend to prefer the second one.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -61,6 +61,75 @@ static inline void popstack(std::vector<valtype>& stack)
+static inline void push8_le(std::vector<valtype>& stack, uint64_t v)
+{
+    auto v_le = htole64(v);
+    valtype vch(reinterpret_cast<unsigned char*>(&v_le), reinterpret_cast<unsigned char*>(&v_le) + 8);
```
emplace_back.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -61,6 +61,75 @@ static inline void popstack(std::vector<valtype>& stack)
+    vchinpAsset.insert(vchinpAsset.begin(), asset.vchCommitment.begin() + 1, asset.vchCommitment.begin() + 33);
+    valtype vchAssetPref;
+    vchAssetPref.insert(vchAssetPref.begin(), asset.vchCommitment.begin(), asset.vchCommitment.begin() + 1);
+    stack.push_back(vchinpAsset);
```
emplace_back.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -61,6 +61,75 @@ static inline void popstack(std::vector<valtype>& stack)
+        // If value is null, explicitly push the explicit prefix 0x01
+        vchValuePref = valtype(1, 1);
+    }
+    stack.push_back(vchinpValue); // if value is null, 0(LE 8) is pushed
```
Keeping in mind that I don't understand move semantics, my understanding from pieter is that we want std::move within these push_backs.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -61,6 +61,75 @@ static inline void popstack(std::vector<valtype>& stack)
+    int witnessversion;
+    valtype witnessprogram;
+    if (scriptPubKey.IsWitnessProgram(witnessversion, witnessprogram)) {
+        stack.push_back(witnessprogram);
```
with the same caveats, std::move here.

sanket1729 commented on src/script/interpreter.cpp:
```diff
@@ -15,6 +15,9 @@
 
 typedef std::vector<unsigned char> valtype;
 
+static_assert(static_cast<uint64_t>(std::numeric_limits<int64_t>::max()) == UINT64_C(0x7FFFFFFFFFFFFFFF));
```
Unfortunately the ' symbol did not work inside the UINT64_C macro

roconnor-blockstream replied:
That's a shame.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -60,6 +64,81 @@ static inline void popstack(std::vector<valtype>& stack)
+static inline void push4_le(std::vector<valtype>& stack, uint32_t v)
+{
+    uint32_t v_le = htole32(v);
+    stack.emplace_back(reinterpret_cast<unsigned char*>(&v_le), reinterpret_cast<unsigned char*>(&v_le) + 4);
```
Replacing 4 with `sizeof(v_le)` is better.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -60,6 +64,81 @@ static inline void popstack(std::vector<valtype>& stack)
+static inline void push8_le(std::vector<valtype>& stack, uint64_t v)
+{
+    uint64_t v_le = htole64(v);
+    stack.emplace_back(reinterpret_cast<unsigned char*>(&v_le), reinterpret_cast<unsigned char*>(&v_le) + 8);
```
Replacing 8 with `sizeof(v_le)` is better.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -14,6 +15,9 @@
 
 typedef std::vector<unsigned char> valtype;
 
+static_assert(static_cast<uint64_t>(std::numeric_limits<int64_t>::max()) == UINT64_C(0x7FFFFFFFFFFFFFFF));
```
Need a comment as to why we are making these asserts.  Something about that these values are consensus critical and used in the Elements 64-bit arithmetic opcodes.

roconnor-blockstream commented on src/script/interpreter.cpp:
```diff
@@ -61,6 +61,68 @@ static inline void popstack(std::vector<valtype>& stack)
+        vchinpValue.assign(value.vchCommitment.begin() + 1, value.vchCommitment.end());
+    }
+    valtype vchValuePref;
+    if (!value.IsNull()) {
```
Let's reverse the arms of this if statement to make it line up with the if statement above.

TBH, I'd somewhat prefer these if statements were merged together.

apoelstra commented:
ack 11d74c37440c6c626bbd9c04801953cf07aa1123 (I know there are more changes coming, just checkpointing my review)

apoelstra commented:
ACK 97597e30c58311f85003c68c7ff9c62f8781d132

apoelstra commented:
ACK 1577ed4e902445e1401a381321634f1a9c843509
